File: documentation-examples-java/src/java/dev/typr/foundations/docs/core/ComposingParsers.java
package dev.typr.foundations.docs.core;

import dev.typr.foundations.And;
import dev.typr.foundations.PgTypes;
import dev.typr.foundations.RowParser;

import java.util.Optional;

@SuppressWarnings("unused")
public class ComposingParsers {
    record ProductRow(Integer id, String name) {}
    record CategoryRow(Integer id, String categoryName) {}

    RowParser<ProductRow> productParser =
        RowParser.<ProductRow>builder()
            .field(PgTypes.int4, ProductRow::id)
            .field(PgTypes.text, ProductRow::name)
            .build(ProductRow::new);

    RowParser<CategoryRow> categoryParser =
        RowParser.<CategoryRow>builder()
            .field(PgTypes.int4, CategoryRow::id)
            .field(PgTypes.text, CategoryRow::categoryName)
            .build(CategoryRow::new);

    //start
    // Inner join — both sides always present
    RowParser<And<ProductRow, CategoryRow>> innerJoined =
        productParser.joined(categoryParser);

    // Left join — right side is Optional (nullable in Kotlin, Option in Scala)
    RowParser<And<ProductRow, Optional<CategoryRow>>> leftJoined =
        productParser.leftJoined(categoryParser);
    //stop
}

File: documentation-examples-scala/src/scala/dev/typr/foundations/docs/core/ComposingParsers.scala
package dev.typr.foundations.docs.core
import dev.typr.foundationssc.*
import dev.typr.foundationssc.data.*

@SuppressWarnings(Array("unused"))
object ComposingParsers:
  case class ProductRow(id: Int, name: String)
  case class CategoryRow(id: Int, categoryName: String)

  val productParser: RowParser[ProductRow] =
    RowParser.builder[ProductRow]()
      .field(PgTypes.int4)(_.id)
      .field(PgTypes.text)(_.name)
      .build(ProductRow.apply)

  val categoryParser: RowParser[CategoryRow] =
    RowParser.builder[CategoryRow]()
      .field(PgTypes.int4)(_.id)
      .field(PgTypes.text)(_.categoryName)
      .build(CategoryRow.apply)

  //start
  // Inner join — both sides always present
  val innerJoined: RowParser[(ProductRow, CategoryRow)] =
    productParser.joined(categoryParser)

  // Left join — right side is Option
  val leftJoined: RowParser[(ProductRow, Option[CategoryRow])] =
    productParser.leftJoined(categoryParser)
  //stop
